reinstate hive health metrics

bimmlerd opened the pull request:
As part of #32020, the hive module health metrics were accidentially dropped. Reinstate them in a new form, and add a test for the functionality.

bimmlerd committed:
hive/health: move healthv2 to hive and rename it

Moves the healthv2 implementation to hive/health to get rid of the
somewhat ugly v2 part of the pkg name, and to give some logical
structure.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
hive/health: readd module health metrics

Somewhere in the move from pkg/hive to cilium/hive we lost the module
health metrics. Add them back in the new form of a query on the status
table, and add a test.

Fixes: 2311f3d781 (treewide: rebase on cilium/hive)

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
hive/health: use slog/logger and minor cleanups

This ensures that logging output of the metrics job and the rest of the
health subsystem are piped through the same logging mechanism. In
addition, unexport what need not be exported, and avoid some amount of
stuttering in "health.HealthTableName", for example.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

nebril reviewed (approved):
Whoops!

dylandreimerink reviewed (approved):
LGTM

bimmlerd committed:
node/manager: test: avoid future race in status

With the introduction of the hive module health metric publisher in a
subsequent path, a precondition of the node manager EmitStatus test will
break: the test assumes the node manager to be the only writer to the
status table. Without this patch, a race occurs.

Since it seems likely that the set of default hive cells will grow (and
that they may want to update their status too) change the test so that
it no longer assumes exclusivity. It now looks specifically at its own
status, and uses revisions to understand when the status changes.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

tommyp1ckles commented on pkg/hive/health/provider.go:
```diff
@@ -125,9 +126,9 @@ func (p *HealthProvider) ForModule(mid cell.FullModuleID) cell.Health {
-			logger.WithField("prefix", i).
-				WithField("deleted", deleted).
-				Debugf("delete health sub-tree")
+			p.logger.Debug("delete health sub-tree",
```
nice cleanup here, thanks!

tommyp1ckles commented on daemon/cmd/daemon_main.go:
```diff
@@ -1654,7 +1654,7 @@ type daemonParams struct {
 	AuthManager          *auth.AuthManager
 	Settings             cellSettings
-	HealthV2Provider     healthTypes.Provider
+	HealthV2Provider     types.Provider
```
having a import named "types" is always a bit annoying in a big file like daemon_main.go 😄

bimmlerd committed:
vendor: upgrade cilium/hive dep

Pulling this bugfix in manually to ensure we don't hit race conditions
after the next commit.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

marseel commented on pkg/node/manager/manager_test.go:
```diff
@@ -854,32 +855,31 @@ func TestNodeManagerEmitStatus(t *testing.T) {
 
 		done := make(chan struct{})
 		reattempt := make(chan struct{})
-		checkStatus := func() ([]types.Status, <-chan struct{}) {
```
Commenting on random line in manager_test.go.

I wonder if it would be simpler if that part:
https://github.com/cilium/cilium/blob/e4e8a15fd783fd0524752090d9fffaa76ff5fa28/pkg/node/manager/manager.go#L362-L367
was pulled into `singleBackgroundLoop` and:

In that case, you wouldn't need to override global `baseBackgroundSyncInterval = 1 * time.Millisecond`. 
After calling `singleBackgroundLoop` you could use `Get` from statedb instead of `GetWatch` which would simplify the test a bit, kind of similar to the current test here: https://github.com/cilium/cilium/blob/e4e8a15fd783fd0524752090d9fffaa76ff5fa28/pkg/node/manager/manager_test.go#L442

bimmlerd replied:
yeah, fair point - I think I'll leave it as is for the merge to go through as it fixes a bug, but took a note to revisit this once I have a bit more cycles

marseel reviewed (approved):
Looks good from vendor point of view, one non-blocking idea for the node manager test.
